# Day 8: Seven Segment Search

[https://adventofcode.com/2021/day/8](https://adventofcode.com/2021/day/8)

## Description

### Part One

You barely reach the safety of the cave when the whale smashes into the cave mouth, collapsing it. Sensors indicate another exit to this cave at a much greater depth, so you have no choice but to press on.

As your submarine slowly makes its way through the cave system, you notice that the four-digit [seven-segment displays](https://en.wikipedia.org/wiki/Seven-segment_display) in your submarine are malfunctioning; <span title="Yes, just the four-digit seven-segment ones. Whole batch must have been faulty.">they must have been damaged</span> during the escape. You'll be in a lot of trouble without them, so you'd better figure out what's wrong.

Each digit of a seven-segment display is rendered by turning on or off any of seven segments named `a` through `g`:

      0:      1:      2:      3:      4:
     aaaa    ....    aaaa    aaaa    ....
    b    c  .    c  .    c  .    c  b    c
    b    c  .    c  .    c  .    c  b    c
     ....    ....    dddd    dddd    dddd
    e    f  .    f  e    .  .    f  .    f
    e    f  .    f  e    .  .    f  .    f
     gggg    ....    gggg    gggg    ....

      5:      6:      7:      8:      9:
     aaaa    aaaa    aaaa    aaaa    aaaa
    b    .  b    .  .    c  b    c  b    c
    b    .  b    .  .    c  b    c  b    c
     dddd    dddd    ....    dddd    dddd
    .    f  e    f  .    f  e    f  .    f
    .    f  e    f  .    f  e    f  .    f
     gggg    gggg    ....    gggg    gggg


So, to render a `1`, only segments `c` and `f` would be turned on; the rest would be off. To render a `7`, only segments `a`, `c`, and `f` would be turned on.

The problem is that the signals which control the segments have been mixed up on each display. The submarine is still trying to display numbers by producing output on signal wires `a` through `g`, but those wires are connected to segments _randomly_. Worse, the wire/segment connections are mixed up separately for each four-digit display! (All of the digits _within_ a display use the same connections, though.)

So, you might know that only signal wires `b` and `g` are turned on, but that doesn't mean _segments_ `b` and `g` are turned on: the only digit that uses two segments is `1`, so it must mean segments `c` and `f` are meant to be on. With just that information, you still can't tell which wire (`b`/`g`) goes to which segment (`c`/`f`). For that, you'll need to collect more information.

For each display, you watch the changing signals for a while, make a note of _all ten unique signal patterns_ you see, and then write down a single _four digit output value_ (your puzzle input). Using the signal patterns, you should be able to work out which pattern corresponds to which digit.

For example, here is what you might see in a single entry in your notes:

    acedgfb cdfbe gcdfa fbcad dab cefabd cdfgeb eafb cagedb ab |
    cdfeb fcadb cdfeb cdbaf

(The entry is wrapped here to two lines so it fits; in your notes, it will all be on a single line.)

Each entry consists of ten _unique signal patterns_, a `|` delimiter, and finally the _four digit output value_. Within an entry, the same wire/segment connections are used (but you don't know what the connections actually are). The unique signal patterns correspond to the ten different ways the submarine tries to render a digit using the current wire/segment connections. Because `7` is the only digit that uses three segments, `dab` in the above example means that to render a `7`, signal lines `d`, `a`, and `b` are on. Because `4` is the only digit that uses four segments, `eafb` means that to render a `4`, signal lines `e`, `a`, `f`, and `b` are on.

Using this information, you should be able to work out which combination of signal wires corresponds to each of the ten digits. Then, you can decode the four digit output value. Unfortunately, in the above example, all of the digits in the output value (`cdfeb fcadb cdfeb cdbaf`) use five segments and are more difficult to deduce.

For now, _focus on the easy digits_. Consider this larger example:

    be cfbegad cbdgef fgaecd cgeb fdcge agebfd fecdb fabcd edb |
    fdgacbe cefdb cefbgd gcbe
    edbfga begcd cbg gc gcadebf fbgde acbgfd abcde gfcbed gfec |
    fcgedb cgb dgebacf gc
    fgaebd cg bdaec gdafb agbcfd gdcbef bgcad gfac gcb cdgabef |
    cg cg fdcagb cbg
    fbegcd cbd adcefb dageb afcb bc aefdc ecdab fgdeca fcdbega |
    efabcd cedba gadfec cb
    aecbfdg fbg gf bafeg dbefa fcge gcbea fcaegb dgceab fcbdga |
    gecf egdcabf bgf bfgea
    fgeab ca afcebg bdacfeg cfaedg gcfdb baec bfadeg bafgc acf |
    gebdcfa ecba ca fadegcb
    dbcfg fgd bdegcaf fgec aegbdf ecdfab fbedc dacgb gdcebf gf |
    cefg dcbef fcge gbcadfe
    bdfegc cbegaf gecbf dfcage bdacg ed bedf ced adcbefg gebcd |
    ed bcgafe cdgba cbgef
    egadfb cdbfeg cegd fecab cgb gbdefca cg fgcdab egfdb bfceg |
    gbdfcae bgc cg cgb
    gcafb gcf dcaebfg ecagb gf abcdeg gaef cafbge fdbac fegbdc |
    fgae cfgab fg bagce


Because the digits `1`, `4`, `7`, and `8` each use a unique number of segments, you should be able to tell which combinations of signals correspond to those digits. Counting _only digits in the output values_ (the part after `|` on each line), in the above example, there are _`26`_ instances of digits that use a unique number of segments (highlighted above).

_In the output values, how many times do digits `1`, `4`, `7`, or `8` appear?_

In [2]:
def part1(filename):
    lines = [line.strip() for line in open(filename) if line.strip()]
    parts = [line.partition(" | ") for line in lines]
    # print(parts)
    data = [part.split() for _, _, part in parts]

    count = sum(sum(len(d) in (2, 3, 4, 7) for d in row) for row in data)
    print(filename, count)


part1("input/day8-test.txt")
part1("input/day8.txt")

input/day8-test.txt 26
input/day8.txt 476


<div class="alert alert-info">Sheesh, this one sounds a bit harder than any we've had so far. I structured this as
part1/part2 functions so less editing to put into a notebook.
</div>

**Your puzzle answer was 476.**

### Part Two

Through a little deduction, you should now be able to determine the remaining digits. Consider again the first example above:

    acedgfb cdfbe gcdfa fbcad dab cefabd cdfgeb eafb cagedb ab |
    cdfeb fcadb cdfeb cdbaf

After some careful analysis, the mapping between signal wires and segments only make sense in the following configuration:

     dddd
    e    a
    e    a
     ffff
    g    b
    g    b
     cccc


So, the unique signal patterns would correspond to the following digits:

*   `acedgfb`: `8`
*   `cdfbe`: `5`
*   `gcdfa`: `2`
*   `fbcad`: `3`
*   `dab`: `7`
*   `cefabd`: `9`
*   `cdfgeb`: `6`
*   `eafb`: `4`
*   `cagedb`: `0`
*   `ab`: `1`

Then, the four digits of the output value can be decoded:

*   `cdfeb`: _`5`_
*   `fcadb`: _`3`_
*   `cdfeb`: _`5`_
*   `cdbaf`: _`3`_

Therefore, the output value for this entry is _`5353`_.

Following this same process for each entry in the second, larger example above, the output value of each entry can be determined:

*   `fdgacbe cefdb cefbgd gcbe`: `8394`
*   `fcgedb cgb dgebacf gc`: `9781`
*   `cg cg fdcagb cbg`: `1197`
*   `efabcd cedba gadfec cb`: `9361`
*   `gecf egdcabf bgf bfgea`: `4873`
*   `gebdcfa ecba ca fadegcb`: `8418`
*   `cefg dcbef fcge gbcadfe`: `4548`
*   `ed bcgafe cdgba cbgef`: `1625`
*   `gbdfcae bgc cg cgb`: `8717`
*   `fgae cfgab fg bagce`: `4315`

Adding all of the output values in this larger example produces _`61229`_.

For each entry, determine all of the wire/segment connections and decode the four-digit output values. _What do you get if you add up all of the output values?_

In [3]:
digits = {
    1: "cf",
    7: "acf",
    4: "bcdf",
    2: "acdeg",
    3: "acdfg",
    5: "abdfg",
    6: "abdefg",
    9: "abcdfg",
    0: "abcefg",
    8: "abcdefg",
}

<div class="alert alert-info">Started with this dict, though I wasn't yet sure how I would use it. In hindsight I
should have just made those strings sets.</div>

In [4]:

def decode(patterns, values):
    known = {}
    for p in patterns:
        if len(p) == 2:
            known[1] = set(p)
        if len(p) == 3:
            known[7] = set(p)
        if len(p) == 4:
            known[4] = set(p)
        if len(p) == 7:
            known[8] = set(p)
    for p in patterns:
        if len(p) == 6:
            if len(set(p).intersection(known[1])) == 1:
                known[6] = set(p)
            elif len(set(p).intersection(known[4])) == 4:
                known[9] = set(p)
            else:
                known[0] = set(p)
    for p in patterns:
        if len(p) == 5:
            if len(set(p).intersection(known[1])) == 2:
                known[3] = set(p)
            elif len(set(p).intersection(known[6])) == 5:
                known[5] = set(p)
            else:
                known[2] = set(p)
    # print(known)
    decoded = []
    for v in values:
        for k in known:
            if known[k] == set(v):
                decoded.append(k)
    # print(decoded)
    assert len(decoded) == 4
    return decoded[0] * 1000 + decoded[1] * 100 + decoded[2] * 10 + decoded[3]

<div class="alert alert-info">Set up a dict of known sets, 1, 7, 4, and 8 are the easy ones,
then the 6 segment values can be extracted from unique intersections with 1 and 4, and the 5 segment
values have unique intersections with 1 and 6. Finally 'known' maps things the wrong way so I have nested
loops when I could have created a segment to digit mapping which would have been easier.
</div>

In [5]:
def part2(filename):
    lines = [line.strip() for line in open(filename) if line.strip()]
    parts = [line.partition(" | ") for line in lines]
    # print(parts)
    data = [(patterns.split(), values.split()) for patterns, _, values in parts]

    real = [decode(p, v) for p, v in data]
    print(filename, sum(real))


print(
    decode(
        "acedgfb cdfbe gcdfa fbcad dab cefabd cdfgeb eafb cagedb ab".split(),
        "cdfeb fcadb cdfeb cdbaf".split(),
    )
)

5353


<div class="alert alert-info">Very similar to part 1 now, parse the input then use the decode function
to get the real values.</div>

In [6]:
part2("input/day8-test.txt")
part2("input/day8.txt")

input/day8-test.txt 61229
input/day8.txt 1011823


**Your puzzle answer was 1011823.**

<div class="alert alert-info">Done.</div>